### DISCUSSION TITLE
Location prefix/file directory support

### DISCUSSION THREAD

**Discussion by jerko-culina** [shrinerb/shrine#699] [Q&A] [answered]
HI, 

Description of problem:
- I have rails application that handles bunch of PDF documents. 
- Application have multiple models that have PDF documents. 
- Some of these models have more than one PDF document (more then one column in database related to PDF document).

Current situation
- Currently I have one PDFUploader that handles/stores all PDF-s
- All PDFs have been stored on the one/same S3 bucket (bucket prefix is let's say 'prod').

What I need:
 -  Way to say each PDF document which directory on S3 PDFs should be stored.

One way to solve it 
- One way to solve that is to have one uploader per PDF document with overridden `generate_location` method.
- Problem is that I have 120 different PDF documents (columns in database)
  - That means i would have to have 120 different PDFUploaders, only difference is location
  - I have to include 120 different uploaders inside app models just because of different location

Question is:
- Is there any better way to solve it?

**Comment by benkoshy** [comment]
> Problem is that I have 120 different PDF documents (columns in database)
> That means i would have to have 120 different PDFUploaders, only difference is location
> I have to include 120 different uploaders inside app models just because of different location

Can you elaborate on this?

120 different columns? Also how are you determining which location that the pdfs are going? Whatever the logic for this is - you can surely extract it out to have just one pdf uploader.

Out of curiousity - why different locations?

(questions asked to avoid XY problems and perhaps a more wholistic solution can be found. ).

**Reply by jerko-culina** [reply]
Sorry, maybe I was not explicit enough. In my application I have 120 different types of PDF documents. Each type looks differently, other words, each one has own design and own generator, etc... 

Example: One model with 3 different PDF documents/types

```ruby
class ModelOne
   field :title
   field ...
   ...
   field :report_pdf_data
   field :pdf_log_data
   field :summary_pdf_data
   
   include ::PdfUploader::Attachment(:report_pdf)
   include ::PdfUploader::Attachment(:pdf_log)
   include ::PdfUploader::Attachment(:summary_pdf)
end
```

In my application I have 120 of them.

> Also how are you determining which location that the pdfs are going? Whatever the logic for this is - you can surely extract it out to have just one pdf uploader.

Mostly on Model they belongs to, and type of document.

For above model example, on S3 I would like to have file and folders architecture something like:
```
/prod/model_one/report_pdfs/file123.pdf
/prod/model_one/pdf_logs/file456.pdf
/prod/model_one/summary_pdfs/file678.pdf
```

> Out of curiousity - why different locations?

Client request (it can easily differentiate, move or copy, etc. them on AWS S3)

**Comment by janko** [comment, marked as answer]
You have access to the current record (`record`) and attachment name (`name`) in `#generate_location`, is that enough to vary the directory per attachment type?

**Reply by jerko-culina** [reply]
That will make it doable, thanks a lot :)

One more question: Is there any consideration/plan/effort to add support for directories. 
From my point of view something like this would be perfect:
```
class ModelOne
   include ::PdfUploader::Attachment(:report_pdf, directory_path: "model_one/report_pdfs")
...
end 
```

In the case I want to create/add plugin that will have that kind of support would it be feasible using plugins?

**Reply by janko** [reply]
Yes, any options passed to `Shrine::Attachment#initialize` are later forwarded to `Shrine::Attacher.new`, so you could create a plugin that stores it in default attacher context:

```rb
# lib/shrine/plugins/directory_path.rb
class Shrine
  module Plugins
    module DirectoryPath
      module AttacherMethods
        def initialize(**, directory_path: nil)
          super(**)
          context[:directory_path] = directory_path
        end
      end
    end

    register_plugin(:directory_path, DirectoryPath)
  end
end
```

Once you load the plugin, the `directory_path` should be automatically available in `#generate_location`:

```rb
class PdfUploader < Shrine
  plugin :directory_path

  def generate_location(io, record: nil, name: nil, directory_path: nil, **)
    directory_path #=> "model_one/report_pdfs"
  end
end
```
